beamer/tests/agent/multiple_chain/conftest.py

```diff
@@ -0,0 +1,217 @@
+    rollup_config,
+    setup_instances,  # pylint:disable=unused-argument
+):
+    chain_contracts = {}
+
+    deployer_local_account = accounts[0]
+
+    l1_w3 = brownie.web3
+    l1_w3.eth.default_account = deployer_local_account.address
+
+    l1_messenger = deploy_contract(l1_w3, ("TestL1Messenger"))
+    transact(l1_messenger.functions.setForwardState(forward_state))
+    resolver = deploy_contract(l1_w3, ("Resolver"))
+    l1_chain_id = brownie.web3.chain_id
+    transact(l1_messenger.functions.addCaller(resolver.address))
+
+    for chain_id, port in rollup_config.items():
+        w3 = Web3(HTTPProvider(f"http://127.0.0.1:{port}", request_kwargs=dict(timeout=5)))
+        while not w3.isConnected():
+            sleep(1)
+        w3.eth.default_account = deployer_local_account.address
+        l2_messenger = deploy_contract(w3, ("TestL2Messenger"))
+        transact(l2_messenger.functions.setForwardState(forward_state))
+        fill_manager = deploy_contract(w3, ("FillManager", l2_messenger.address))
+        transact(fill_manager.functions.setResolver(resolver.address))
+        request_manager = deploy_contract(
+            w3,
+            (
+                "RequestManager",
+                claim_stake,
+                claim_request_extension,
+                claim_period,
+                challenge_period_extension,
+            ),
+        )
+        transact(fill_manager.functions.addAllowedLp(_AGENT_ACCOUNT.address))
+        transact(request_manager.functions.addAllowedLp(_AGENT_ACCOUNT.address))
+
+        transact(l2_messenger.functions.addCaller(fill_manager.address))
+        transact(
+            resolver.functions.addCaller(chain_id, l2_messenger.address, l1_messenger.address)
+        )
+        transact(
+            request_manager.functions.addCaller(
+                l1_chain_id, l1_messenger.address, l2_messenger.address
+            )
+        )
+        transact(
+            resolver.functions.addRequestManager(
+                chain_id, request_manager.address, l1_messenger.address
+            )
+        )
+        for _chain_id in rollup_config:
+            if _chain_id == chain_id:
+                continue
+            transact(request_manager.functions.setFinalityPeriod(_chain_id, finality_period))
+        token = deploy_contract(w3, ("MintableToken", int(1e18)))
+        transact(
+            request_manager.functions.updateToken(
+                token.address, int(10_000e18), int(5e18), 1_000, 0
+            )
+        )
+        rollup_info = _RollupInfo(
+            chain_id=chain_id,
+            rpc_url=URL(f"http://127.0.0.1:{port}"),
+        )
+        chain_contracts[rollup_info] = _RollupContracts(
+            l1_messenger=l1_messenger,
+            l2_messenger=l2_messenger,
+            resolver=resolver,
+            fill_manager=fill_manager,
+            request_manager=request_manager,
+            token=token,
+        )
+    return chain_contracts
```

Comment:
It looks very similar to the deploy function in the conftest, one dir up. could you  outsource the similiar code into a utility function?

Reply:
it is not actually, this one is deploying things by looking at the rpc and returning web3 contracts while the other one directly using brownie rpc only and returning Brownie type ContractInstances

Reply:
and we have token here as an extra